### ISSUE TITLE
Add new RPC-style API to create connections

### ISSUE THREAD

**Issue by seancolsen** [centerofci/mathesar#3293]
This API is to allow the front end to submit the New Database Connection form, as described in these [specs](https://wiki.mathesar.org/design/specs/new-db-connection-form/).

The endpoint should be `/api/ui/v0/connections/`

Note that this is under the `ui` namespace, not `db`.

Requests should be submitted via POST.

The POST request body, as described via TypeScript, is:

```ts
type SampleDataSchemaIdentifier = 'library_management' | 'movie_collection';

type ConnectionToInternalDatabase = {
  connection_type: 'internal_database';
};
type ConnectionToUserDatabase = {
  connection_type: 'user_database';
  /** The Django ID from the connection object */
  id: number;
};
type Connection = ConnectionToInternalDatabase | ConnectionToUserDatabase;

type CredentialsCopied = {
  strategy: 'copy_from_known_connection';
  connection: ConnectionReference;
};
type CredentialsFromScratch = {
  strategy: 'from_scratch';
  user: string;
  password: string;
  host: string;
  port: string;
};
type CredentialsFromScratch = {
  strategy: 'with_new_user';
  create_user_via: Connection;
  new_user: string;
  new_password: string;
  host: string;
  port: string;
};

type Credentials =
  | CredentialsCopied
  | CredentialsFromScratch
  | CredentialsWithNewUser;

type RequestBody = {
  credentials: Credentials;
  database_name: string;
  create_database: boolean;
  sample_data: SampleDataSchemaIdentifier[];
  nickname: string;
};
```

Examples:

- Case A

    ```json
    {
      "credentials": {
        "strategy": "copy_from_known_connection",
        "connection": {
          "connection_type": "internal_database"
        }
      },
      "database_name": "mathesar",
      "create_database": true,
      "sample_data": [],
      "nickname": "My new connection"
    }
    ```

- Case B

    ```json
    {
      "credentials": {
        "strategy": "from_scratch",
        "user": "banana",
        "password": "riYwZ1xQ81RW28aU75k",
        "host": "example.com",
        "port": "5432"
      },
      "database_name": "mathesar",
      "create_database": true,
      "sample_data": [],
      "nickname": "My new connection"
    }
    ```

- Case C

    ```json
    {
      "credentials": {
        "strategy": "with_new_user",
        "create_user_via": {
          "connection_type": "user_database",
          "id": 81
        },
        "user": "banana",
        "password": "riYwZ1xQ81RW28aU75k",
        "host": "example.com",
        "port": "5432"
      },
      "database_name": "mathesar",
      "create_database": true,
      "sample_data": [],
      "nickname": "My new connection"
    }
    ```

**Comment by rajesh604** [comment]
@seancolsen can i work on this issue

**Comment by seancolsen** [comment]
No @rajesh604 this issue is restricted to maintainers

**Comment by rajesh604** [comment]
Can you tell me what are the issues I can pick as a contributor.

**Comment by seancolsen** [comment]
@rajesh604 

> what are the issues I can pick as a contributor

1. Read our [Contributor Guide](https://github.com/centerofci/mathesar/blob/develop/CONTRIBUTING.md).
2. See the "Find an issue to work on" step.
3. Make sure to read the rest of the steps too, along with our [Developer Guide](https://github.com/centerofci/mathesar/blob/develop/DEVELOPER_GUIDE.md).
4. If you still have questions, come ask questions in [Matrix](https://wiki.mathesar.org/community/matrix/)

**Comment by rajesh604** [comment]
> @rajesh604
> 
> > what are the issues I can pick as a contributor
> 
> 1. Read our [Contributor Guide](https://github.com/centerofci/mathesar/blob/develop/CONTRIBUTING.md).
> 2. See the "Find an issue to work on" step.
> 3. Make sure to read the rest of the steps too, along with our [Developer Guide](https://github.com/centerofci/mathesar/blob/develop/DEVELOPER_GUIDE.md).
> 4. If you still have questions, come ask questions in [Matrix](https://wiki.mathesar.org/community/matrix/)

okay

**Comment by seancolsen** [comment]
@mathemancer that all makes sense and I'll defer to your opinions on those questions.

**Comment by mathemancer** [comment]
@seancolsen What would you think of splitting into 3 endpoints under `api/ui/v0/connections/`, one for each strategy?
- Multiple dispatch is a pain while we're simulating RPC endpoints with Django Rest Framework
- Multiple dispatch needs to be handled on the FE anyway, since it has to fill that field
  - Thus, we either handle multiple dispatch _only_ on the front end, or in both the FE and BE (and keep the types aligned between the FE and BE moving forward)
- Cleaner, when thinking of the endpoints as functions (subjective; I don't like the nested value against which we need to validate the whole submission)

**Comment by seancolsen** [comment]
@mathemancer and I chatted about this on Matrix. We'll be moving forward with his "three endpoints" approach and the `credentials.strategy` property will be removed.

**Comment by mathemancer** [comment]
@seancolsen For Case C, I'm going to remove the "host" and "port" arguments; they're (at best) redundant, since the user and DB must be created on the DB server accessed by the 'via' connection. In case they're _not_ redundant, then creating that user and DB via the given connection would be impossible anyway.

**Comment by seancolsen** [comment]
@mathemancer 

> For Case C, I'm going to remove the "host" and "port" keys

Right. That makes sense. I don't think this change will require any frontend adjustments. Go for it.
